zerver/lib/markdown/__init__.py

```diff
@@ -575,15 +575,32 @@ def get_tweet_id(url: str) -> Optional[str]:
     return tweet_id_match.group("tweetid")
 
 
-class InlineHttpsProcessor(markdown.treeprocessors.Treeprocessor):
+class InlineImageProcessor(markdown.treeprocessors.Treeprocessor):
+    """
+    Rewrite inline img tags to serve external content via Camo.
+
+    This rewrites all images, except ones that are served from the current
+    realm or global STATIC_URL. This is to ensure that each realm only loads
+    images that are hosted on that realm or by the global installation,
+    avoiding information leakage to external domains or between realms. We need
+    to disable proxying of images hosted on the same realm, because otherwise
+    we will break images in /user_uploads/, which require authorization to
+    view.
+    """
+
     def run(self, root: Element) -> None:
         # Get all URLs from the blob
         found_imgs = walk_tree(root, lambda e: e if e.tag == "img" else None)
         for img in found_imgs:
             url = img.get("src")
             assert url is not None
-            if urllib.parse.urlsplit(url).scheme != "http":
-                # Don't rewrite images on our own site (e.g. emoji).
+            split_url = urllib.parse.urlsplit(url)
+            if (
+                url.startswith(settings.STATIC_URL)
+                or url.startswith(settings.EXTERNAL_URI_SCHEME + self.md.zulip_realm.host)
```

Comment:
This middle one would be better written with e.g. `split_url.netloc == self.md.zulip_realm.host and split_url.scheme +"://" == self.EXTERNAL_URL_SCHEME`?  

`startswith` is not a correct way to check URLs, because e.g. `https://zulip.example.comic` would match this test.

See also why I wanted this as a function: https://github.com/zulip/zulip/pull/17755/commits/1c012909c29ce7a4a99d49e607b9f657e8cd2509#r600822787 -- this is security-sensitive code so we should put in a common library used to hold such code, and make it easy to write short tests for it (in addition to the ones you have for full markdown system).

Reply:
Fixed, and moved into a function in `zerver/lib/thumbnail.py`. I'm not sure that that's the best place for it, let me know if it should be elsewhere.